src/telemetry/Manager.cc

```diff
@@ -16,20 +17,40 @@
 
 #include "zeek/3rdparty/doctest.h"
 #include "zeek/ID.h"
+#include "zeek/RunState.h"
 #include "zeek/ZeekString.h"
 #include "zeek/broker/Manager.h"
+#include "zeek/iosource/Manager.h"
 #include "zeek/telemetry/ProcessStats.h"
 #include "zeek/telemetry/Timer.h"
 #include "zeek/telemetry/telemetry.bif.h"
 #include "zeek/threading/formatters/detail/json.h"
 
 namespace zeek::telemetry {
 
-Manager::Manager() { prometheus_registry = std::make_shared<prometheus::Registry>(); }
+/**
+ * Prometheus Collectable interface used to insert Zeek callback processing
+ * before the Prometheus registry's collection of metric data.
+ */
+class ZeekCollectable : public prometheus::Collectable {
+public:
+    std::vector<prometheus::MetricFamily> Collect() const override {
+        telemetry_mgr->WaitForPrometheusCallbacks();
+        return {};
+    }
+};
+
+Manager::Manager() {
+    prometheus_registry = std::make_shared<prometheus::Registry>();
+    collector_flare = std::make_unique<zeek::detail::Flare>();
+}
 
-// This can't be defined as =default because of the use of unique_ptr with a forward-declared type
-// in Manager.h
-Manager::~Manager() {}
+Manager::~Manager() {
+    // Shut down the exposer first of all so we stop getting requests for
+    // data. This keeps us from getting a request on another thread while
+    // we're shutting down.
+    prometheus_exposer.reset();
```

Comment:
Suspect if currently a request is processed, doing this is still unsafe, but don't have an immediate answer.

Reply:
I dug down into the civetweb code and their `mg_stop` method correctly handles this case. Since deleting the exposer is effectively just deleting the civetweb handler, it should be fine.